### ISSUE TITLE
Replace "flycheck-compile" with a "flycheck-debug-checker" command

### ISSUE THREAD

**Issue by swsnr** [flycheck/flycheck#854]
`flycheck-compile` is intended as a debugging tool, but it does a poor job at this:
- The name is absolutely misleading :sunglasses: 
- It runs syntax checkers differently from Flycheck, causing escaping issues (see #852)
- It doesn't help with interpreting the output.  Specifically it doesn't show what errors Flycheck would parse from these errors.

I would like to remove `flycheck-compile`, and introduce a new command `flycheck-debug-checker` instead which is really targeted at inspecting and debugging a syntax checker output.  The command should
- [ ] have a nice name (we can find something better than `flycheck-debug-checker` hopefully)
- [ ] run the checker like Flycheck does, but keep the raw output
- [ ] pop up a buffer that includes the raw output and show the errors parsed from that output
- [ ] be documented in our contributing guide as general debugging tool
- [ ] be documented in the manual chapter about defining syntax checkers
## Suggested layout:

```
Run checker `foo' on buffer bar.foo, parsed errors:

<stdin>: line 1: column 2: level warning: A bad error (id: xxxx)
…

From output:
…

Command: foo --lint --standard-input
```

This layout ensures that in case Flycheck fails to parse errors the output comes right at the beginning of the buffer, to minimise scrolling.

**Comment by cpitclaudel** [comment]
Yes, I think that would be nice :) Sorry for pointing users to that command a bit too quickly!

**Comment by lunaryorn** [comment]
@cpitclaudel That's totally fine.  It's there for debugging, and we should tell users to user it for this purpose.

The confusion in #852 was a very unfortunate coincidence.  We'd believe that `shell-quote-argument` has its flaws?! :flushed:

**Comment by cpitclaudel** [comment]
:)

If we can minimize that quoting example, I'll be happy to file a bug. It's surprising to say the least.

**Comment by cmarqu** [comment]
What I also can imagine speeding up the debugging process for a new checker setup is to have a (multiline) string with a copied example output of the checker (info, warning, error one after the other) and a command to run this string through the matcher function. 
This add implicit documentation and could even become part of a regression test.
When e.g. a checker stops working or doesn't find a specific error output, then that makes it easier to compare the working example with your specific failing example.

**Comment by lunaryorn** [comment]
@cmarqu I'm not sure whether I understand you.  Are you suggesting that we add an example output to every defined syntax checker, e.g. via an `:example` keyword?

**Comment by cmarqu** [comment]
No, rather an example for the *input* of a checker, i.e. one error line of the output of the tool (say, pylint).

**Comment by lunaryorn** [comment]
@cmarqu Yep, that's what I meant—I was referring to the output of the underlying tool.  Just to clarify, you'd like to see a syntax checker definition such as:

```
(flycheck-define-checker pylint
  :command (pylint …)
  :example "3:10: warning: Undefined function foo"
  …)
```

**Comment by cmarqu** [comment]
Ah, yes, sorry for the misunderstanding, our example is what I had in mind.
You'd maybe want to separate that :example-info, :example-warning, :example-error and have them accept lists of strings (sometimes tools produce output both with and without line numbers etc.), but I'll leave the details to you.

**Comment by lunaryorn** [comment]
@cmarqu Honestly, I don't think that this is a good idea.  What purpose would these examples serve?  What problems would they solve that'd balance the enormous effort of adding and maintaining them?

**Comment by lunaryorn** [comment]
What follows is a summary of a discussion @cmarqu and me had about the shape of this feature [on Gitter](https://gitter.im/flycheck/flycheck?at=56a901dbeaf741c118d53b0c).  Specifically, @cmarqu raised concerns about the need for excessive scrolling in a debugging buffer, which we would address by ordering the buffer appropriately to minimise the amount of scrolling necessary.  For this purpose we'd need to ensure to prominently place the output particularly if the command fails to parse any checkers.

I've updated the original post with a suggested layout.

**Comment by swsnr** [comment]
This issue wasn't updated in a long time, and I am no longer interested, so I'm closing this issue to clean up my issues list.
